### PR TITLE
Allow value broadcasting in distributions.Distribution

### PR DESCRIPTION
**Pull request by fritzo**
This relaxes `Distribution._validate_log_prob_arg()` to allow broadcasting. This came up in a simple use case:
```py
p = Normal(Variable(torch.Tensor([0.0])),
           Variable(torch.Tensor([1.0])))
x = torch.arange(-2,2,0.01)
pdf = torch.exp(p.log_prob(x))  # Fails before this PR, works after.
pyplot.plot(x.numpy(), pdf.numpy())
```
This also removes two unused imports of `expand_n` from files in torch.distributions.

[Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit)

### CODE REVIEW

**Comment by fritzo** [comment]
@neerajprad Does this seem to permissive? I think this won't be as much of an issue once we have scalar support.

File: torch/distributions/distribution.py [hunk 1]
```diff
@@ -96,7 +96,9 @@ def _validate_log_prob_arg(self, value):
         """
         if not (torch.is_tensor(value) or isinstance(value, Variable)):
             raise ValueError('The value argument to log_prob must be a Tensor or Variable instance.')
-        batch_dim_start = len(value.size()) - len(self._batch_shape) - len(self._event_shape)
-        if value.size()[batch_dim_start:] != self._batch_shape + self._event_shape:
-            raise ValueError('The right-most size of value must match: {}.'.
-                             format(self._batch_shape + self._event_shape))
+        expected_shape = self._batch_shape + self._event_shape
+        actual_shape = value.size()
+        for i, j in zip(reversed(actual_shape), reversed(expected_shape)):
```
Comment (apaszke) [inline torch/distributions/distribution.py]: You could also try using inferSize in a try except I think

**Comment by neerajprad** [comment]
> @neerajprad Does this seem to permissive? I think this won't be as much of an issue once we have scalar support.

Interesting. This won't be an issue once we have scalar support. The current status is this:
 - support scalars as arguments to univariate distributions, but preserve the return type as tensors. 
 - `log_prob` should throw an error if the value to be scored could not have been generated by any possible invocation of `sample` with the same parameters.
 - no broadcasting for the batch and event dimensions. Specially for the event dimension, we get into this weird issue where expanding the parameters might result in invalid parameterization. 

What we are basically doing here is relaxing the requirement for batch dims to line up (changing the second and third point above). e.g. even this should be possible now:
```
        p = Normal(Variable(torch.Tensor([[0.0], [0.0]])),
                   Variable(torch.Tensor([[1.0], [1.0]])))
        x = Variable(torch.arange(-2,2,0.01))
        pdf = torch.exp(p.log_prob(x))  # returns score of size (2, 401)
```

For the most part, this should be fine, but will add some complexity in the longer term. Also doing this just for this one use case seems a bit overkill? 

Another option to consider would be to preserve the current batch semantics and instead requiring the user to use `p = Normal(0.0, 1.0)` to be able to score batches like this. We will need to upcast to Variable instead of Tensor internally, so that we are able to differentiate. This will require no changes to the client side code once we have scalar support, and preserve our batching semantics. Otherwise, the above scoring operation should throw an error once we have scalar support (as it does in Tensorflow). Relaxing the requirement for batch dims to line up is fine too if this is a common use case, but I would need to think a bit more about other side effects. What do you think?

**Comment by fritzo** [comment]
@neerajprad What if we allow broadcasting over `batch_shape` but require `event_shape` to match exactly? That would be enough to resolve this use case.

**Comment by neerajprad** [comment]
> What if we allow broadcasting over batch_shape but require event_shape to match exactly? That would be enough to resolve this use case.

That's what the change in this PR does to an extent (we can check for event dims explicitly, and simplify the logic). It will resolve this issue, but relaxes broadcasting requirements more generally to handle this one case:
 - we are allowing the distribution instance to score samples that could not have been generated through `sample`, only to handle the scalar case. Otherwise, in practice, we shouldn't need this broadcasting.
 - the output from `log_prob` will be broadcasted, and the result can be of a very different size than the value to be scored. The dimensions won't simply be `value.Size()[:-event_dim]`.

To be fair, none of these seem like that big a deal to me at this point. I would prefer to stay close to TensorFlow semantics, in the absence of strong reasons not to do so. Not having scalar support is a valid reason to deviate from that, but can we handle this with minimal changes to the semantics? Can we instead require clients to use scalar parameters to score such shapes instead? This will probably introduce incompatibilities on the Pyro side, which we will have to handle. :( Or we could be more specific, and allow this expansion of the parameters only for `batch_shape = (1,) and event_shape=()`. When we have scalar support, we could remove this relaxation (the change won't be backwards compatible).

**Comment by fritzo** [comment]
@neerajprad You've convinced me that values should match `event_shape` exactly. I think there is a strong case for allowing broadcasting-up of `batch_shape` of values to match parameters (which is very useful in mixture models: scoring a single datum over many parameters). There's less of a case for broadcasting-down `batch_shape` of many values to parameters of 1 value, but this behavior would allow us to move past scalar issues.

> require clients to use scalar parameters to score such shapes

This doesn't work currently because I usually need a Variables, and Variables must have at least one dimension. @neerajprad let's chat tomorrow.

**Comment by neerajprad** [comment]
> (which is very useful in mixture models: scoring a single datum over many parameters)

Yeah, broadcasting up of values to match distribution parameters could be useful in this context. Sure, let's discuss tomorrow. I will also think about it a bit more.

**Comment by fritzo** [comment]
Ok, I've updated to check (1) exact matching of event_shape and (2) broadcastable to batch_shape.

@apaszke I tried to use `torch._C._infer_size()` as you suggest but it does not yet support empty sizes. We can switch to that once empty sizes are supported.

File: torch/distributions/distribution.py [hunk 2]
```diff
@@ -96,7 +96,15 @@ def _validate_log_prob_arg(self, value):
...
-                             format(self._batch_shape + self._event_shape))
+
+        event_dim_start = len(value.size()) - len(self._event_shape)
+        if value.size()[event_dim_start:] != self._event_shape:
+            raise ValueError('The right-most size of value must match event_shape: {} vs {}.'.
+                             format(value.size(), self._event_shape))
+
+        actual_shape = value.size()
+        expected_shape = self._batch_shape + self._event_shape
+        for i, j in zip(reversed(actual_shape), reversed(expected_shape)):
```
Comment (neerajprad) [inline torch/distributions/distribution.py]: It may be cheaper to do
```
    if self._batch_shape and value.size()[:event_dim_start]:
        try:
            torch._C._infer_size(value.size()[:event_dim_start], self._batch_shape)
        except RuntimeError:
            raise ValueError('Value is not broadcastable with batch_shape+event_shape: {} vs {}.'.
                             format(actual_shape, expected_shape))
```

Just to make sure, this broadcasting will allow both upcasting of distribution parameters as well as upcasting of values:

e.g.
```
        p = Normal(Variable(torch.Tensor([[0.0], [1.0]])),
                   Variable(torch.Tensor([[1.0], [1.0]])))    # size (2, 1)
        x = Variable(torch.Tensor([0.0, 1.0]))    # size (2,)
        pdf = torch.exp(p.log_prob(x))     # size (2, 2)
```

**Review by neerajprad — APPROVED** [review]
Discussed with @fritzo and thought more about it - I do not see any issues with relaxing the constraint on batch dimension matching. Since it will be useful for us to handle the scalar case (and mixture models), lets go ahead with this. Lets use `torch._C._infer_size` to check broadcasting for the batch dim.

**Review by apaszke — APPROVED** [review]
LGTM. @neerajprad you still have this one comment. Let me know if I should go ahead and merge this, or if you guys still want to update that

**Comment by apaszke** [comment]
Actually I see that you approved it later, so it should be ok. Merging this
